h2o-core/src/main/java/hex/ModelBuilder.java

```diff
@@ -2143,7 +2151,16 @@ IcedHashMap<Key, String> getToDelete(boolean expensive) {
     }
 
     void cleanUp() {
-      FrameUtils.cleanUp(_toDelete);
+      // converting Workspace-tracked keys to Scope-tracked keys
+      // much safer than strictly removing everything as frame like training/validation frames are protected in Scope.
+      Key[] tracked = _toDelete.keySet().toArray(new Key[0]);
+      for (Key k: tracked) {
+        Value v = DKV.get(k);
+        if (v==null) continue;
+        if (v.isFrame()) Scope.track(v.get(Frame.class));
+        else if (v.isVec()) Scope.track(v.get(Vec.class));
+        else Scope.track_generic(v.get(Keyed.class));
+      }
```

Comment:
btw, I tend to think that the` _toDelete` map attached to `Model` to play the same role as the workspace in `ModelBuilder` is kinda trash.
It means that each time we score a Frame against the model, we keep a reference to some of its Vecs (the categorical encoded ones most likely), for the entire lifecycle of the model. This is extremely ugly imo.